pact/Pact/Core/Serialise/CBOR_V1.hs

```diff
@@ -76,8 +90,67 @@ encodeNamespace = toStrictByteString . encode
 
-encodeRowData :: RowData -> ByteString
-encodeRowData = toStrictByteString . encode
+-- TODO: Do we want a gas log entry here???
+-- most likely not, maybe just pre/post?
+chargeGasMSerialize :: i -> MilliGas -> GasM (PactError i) ()
+chargeGasMSerialize info amount = do
+  let fakeStackFrame = [] -- TODO: Greg: use real stack frame?
+  (GasMEnv gasRef mgl@(MilliGasLimit gasLimit)) <- ask
+  !currGas <- liftIO $ readIORef gasRef
+  let !used = currGas <> amount
+  liftIO (writeIORef gasRef used)
+  when (used > gasLimit) $ throwError (PEExecutionError (GasExceeded mgl used) fakeStackFrame info)
+
+encodeRowData :: i -> RowData -> GasM (PactError i) ByteString
+encodeRowData info rd = do
+  gasSerializeRowData info rd
+  pure . toStrictByteString $ encode rd
+
+gasSerializeRowData :: forall i. i -> RowData -> GasM (PactError i) ()
+gasSerializeRowData info (RowData fields) = do
+
+  -- Charge for keys
+  -- According to our benchmarking (see `Bench.hs`), it costs
+  -- 183 + (250 * nKiloCharacters in keys) to serialize the keys in a RowData.
+  -- We pad this for safety.
+  let offsetCostMilligas = 200
+  let perKiloCharacterCostMilligas = 400
+  let nKiloCharacters = fromIntegral (sum $ Text.length . _field <$> Map.keys fields) `div` 1000
+  chargeGasMSerialize info $ MilliGas $ offsetCostMilligas + perKiloCharacterCostMilligas * nKiloCharacters 
+
+  -- Charge for values
+  traverse_ gasSerializePactValue fields
+
+  where
+
+    gasSerializePactValue :: PactValue -> GasM (PactError i) ()
+    gasSerializePactValue = \case
+      PLiteral l -> gasSerializeLiteral l
+      PList vs -> do
+        chargeGasMSerialize info $ MilliGas 1000
+        traverse_ gasSerializePactValue vs
+      PGuard _ -> chargeGasMSerialize info $ MilliGas 1000
+      PModRef _ -> chargeGasMSerialize info $ MilliGas 1000
+      PObject o -> do
+        chargeGasMSerialize info $ MilliGas $ (1000 *) $ sum $ fromIntegral . Text.length . _field <$> Map.keys o
+        traverse_ gasSerializePactValue o
+      PCapToken (CapToken name args) -> do
+        chargeGasMSerialize info $ MilliGas $ 1 + 69 * fromIntegral (Text.length (renderText name))
+        traverse_ gasSerializePactValue args
+      PTime _ -> chargeGasMSerialize info $ MilliGas 184
+
+    gasSerializeLiteral = \case
+      LString s ->
+        -- See the analysis in `Bench.hs` - `pact-string-2` for details.
+        chargeGasMSerialize info $ MilliGas $ 1 + 69 * fromIntegral (Text.length s) `div` 1000
+      LInteger i ->
+        -- See the analysis in `Bench.hs` - `pact-ineger-2` for details.
+        chargeGasMSerialize info $ MilliGas $ 2 * fromIntegral (I# (IntLog.integerLogBase# 10 (abs i)))
+      LDecimal d ->
+        chargeGasMSerialize info $ MilliGas $ 59 + 2 * fromIntegral (I# (IntLog.integerLogBase# 10 (decimalMantissa d)))
+      LBool _ -> chargeGasMSerialize info $ MilliGas 52
+      LUnit -> chargeGasMSerialize info $ MilliGas 51
```

Comment:
I see what you mean about the constants. 

Maybe a `Pact.Gas.Constants` is necessary. There's definitely quite a lot here in this file.